django/db/migrations/state.py

```diff
@@ -107,6 +112,176 @@ def remove_model(self, app_label, model_name):
+        # sufficient if the new field is (#27737).
+        # Delay rendering of relationships if it's not a relational field and
+        # not referenced by a foreign key.
+        delay = (
+            not field.is_relation and
+            not field_is_referenced(
+                self, (app_label, operation.model_name_lower), (operation.name, field),
+            )
+        )
+        self.reload_model(app_label, operation.model_name_lower, delay=delay)
+
+    def rename_field(self, app_label, operation):
+        model_state = self.models[app_label, operation.model_name_lower]
+        # Rename the field
+        fields = model_state.fields
+        try:
+            found = fields.pop(operation.old_name)
+        except KeyError:
+            raise FieldDoesNotExist(
+                "%s.%s has no field named '%s'" % (app_label, operation.model_name, operation.old_name)
+            )
+        fields[operation.new_name] = found
+        for field in fields.values():
+            # Fix from_fields to refer to the new field.
+            from_fields = getattr(field, 'from_fields', None)
+            if from_fields:
+                field.from_fields = tuple([
+                    operation.new_name if from_field_name == operation.old_name else from_field_name
+                    for from_field_name in from_fields
+                ])
+        # Fix index/unique_together to refer to the new field
+        options = model_state.options
+        for option in ('index_together', 'unique_together'):
+            if option in options:
+                options[option] = [
+                    [operation.new_name if n == operation.old_name else n for n in together]
+                    for together in options[option]
+                ]
+        # Fix to_fields to refer to the new field.
+        delay = True
+        references = get_references(
+            self, (app_label, operation.model_name_lower), (operation.old_name, found),
+        )
+        for *_, field, reference in references:
+            delay = False
+            if reference.to:
+                remote_field, to_fields = reference.to
+                if getattr(remote_field, 'field_name', None) == operation.old_name:
+                    remote_field.field_name = operation.new_name
+                if to_fields:
+                    field.to_fields = tuple([
+                        operation.new_name if to_field_name == operation.old_name else to_field_name
+                        for to_field_name in to_fields
+                    ])
+        self.reload_model(app_label, operation.model_name_lower, delay=delay)
+
+    def separate_database_and_state(self, app_label, operation):
+        for state_operation in operation.state_operations:
+            state_operation.state_forwards(app_label, self)
+
+    def run_sql(self, app_label, operation):
+        for state_operation in operation.state_operations:
+            state_operation.state_forwards(app_label, self)
+
```

Comment:
These don't make sense, the state logic should not know about `Operation` instances at all.